python/plugins/processing/algs/gdal/Buffer.py

```diff
@@ -100,9 +100,12 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         if source is None:
             raise QgsProcessingException(self.invalidSourceError(parameters, self.INPUT))
         fields = source.fields()
-        source_details = self.getOgrCompatibleSource(self.INPUT, parameters, context, feedback, executing)
-
-        geometry = self.parameterAsString(parameters, self.GEOMETRY, context)
+        source_details = self.getOgrCompatibleSource(self.INPUT, parameters, context, feedback, executing=True)
+        # QgsVectorFileWriter defaults to geometry column called geom when exporting virtual, memory and grass layer
+        if source_details.layer_name == 'INPUT' and 'INPUT.gpkg' in source_details.connection_string:
+            geometry = parameters[self.GEOMETRY] = 'geom'
+        else:
+            geometry = self.parameterAsString(parameters, self.GEOMETRY, context)
```

Comment:
Instead of handling this here, I'd rather see:

1. An optional "geometry_column_name" field added to the GdalConnectionDetails class. (defaulting to None)
2. GdalAlgorithm.getOgrCompatibleSource updated so that the geometry_column_name field is set to "geom" when the source provider is the virtual provider
3. All the GDAL algorithms which have a geometry column name updated to ignore this value and use the geometry_column_name from the GdalConnectionDetails object if it's set

How's that sound to you?